docs/JAX_for_LLM_pretraining.md

```diff
@@ -0,0 +1,448 @@
+<!-- #endregion -->
+
+```python colab={"base_uri": "https://localhost:8080/"} id="6zMsOIc7ouCO" outputId="b16b0781-d1bd-456a-9757-533562f581df"
+!pip install jax-ai-stack
+!pip install -U tiktoken grain matplotlib
+
+import jax
+jax.devices()
+```
+
+<!-- #region id="OHzJ_bokoovZ" -->
+Get the [TinyStories dataset from Hugging Face](https://huggingface.co/datasets/roneneldan/TinyStories). We only use the training split.
+<!-- #endregion -->
+
+```python colab={"base_uri": "https://localhost:8080/"} id="wUjQsgQEmI1N" outputId="431fc1e1-8f7b-4062-ec02-fbbbde4a90c4"
+!wget https://huggingface.co/datasets/roneneldan/TinyStories/resolve/main/TinyStories-train.txt?download=true -O TinyStories-train.txt
+```
+
+<!-- #region id="sKE2uUafLobI" -->
+Take care of the imports.
+<!-- #endregion -->
+
+```python id="MKYFNOhdLq98"
+import jax
+import jax.numpy as jnp
+import flax.nnx as nnx
+import optax, orbax
+from typing import Any
+import os
+from collections import Counter
+from dataclasses import dataclass
+import grain.python as pygrain
+from jax.experimental import mesh_utils
+from jax.sharding import Mesh, PartitionSpec as P, NamedSharding
+import pandas as pd
+import tiktoken, time
+```
+
+<!-- #region id="rPyt7MV6prz1" -->
+## Build the model
+
+One of the biggest advantages of JAX is how easy it is to enable parallelism. To demonstrate this, we are going to use 4-way data parallel and 2-way tensor parallel. Tensor parallelism is one kind of model parallelism, which shards model tensors; there are other kinds of model parallelism, which we won't cover in this tutorial.
+
+As a background, data parallel means splitting a batch of training data into multiple parts (this is called sharding); this way you can use bigger batch sizes to accelerate training, if you have multiple devices that can run in parallel. On the other hand, you can shard not just the training data. Sometimes your model is so big that the model parameters don't fit on a single accelerator. In this case, tensor parallel helps splitting the parameter tensors within a model onto multiple accelerators so that the model can actually run. Both approaches can take advantage of modern accelerators. For example, TPU v2 on the free Colab tier offers 4 chips, each of which has 2 TPU cores. So this architeture works well with 4-way data parallel and 2-way tensor parallel.
+
+To get a detailed understanding of how JAX automatic parallelism works, please refer to this [JAX tutorial](https://jax.readthedocs.io/en/latest/notebooks/Distributed_arrays_and_automatic_parallelization.html#way-batch-data-parallelism-and-2-way-model-tensor-parallelism). In our case to leverage parallelism, we first need to define a `Mesh`, which declares the TPU resources with 2 axes: `batch` axis as 4 and `model` axis as 2, which maps to the TPU v2 cores. Here, the `model` axis enables the tensor parallel for us.
```

Comment:
Let's link to the intro parallelism doc instead: https://jax.readthedocs.io/en/latest/sharded-computation.html It's a somewhat gentler introduction.

Reply:
done